jax/experimental/jax2tf/jax2tf.py

```diff
@@ -126,6 +127,119 @@ def add_unit(v: TfValOrUnit, aval: core.AbstractValue):
     return core.unit if aval is core.abstract_unit else v
   return util.safe_map(add_unit, vals, avals)
 
+
+class SingletonDecorator:
+  def __init__(self, cls):
+    self.cls = cls
+    self.instance = None
+
+  def __call__(self, *args, **kwargs):
+    if not self.instance:
+      self.instance = self.cls(*args, **kwargs)
+    return self.instance
+
+PrimitiveName = str
+ErrorType = str
+ErrorString = str
+Devices = Tuple[str, ...]
+Limitation = Tuple[PrimitiveName, ErrorType, ErrorString, Devices]
+
+def categorize(prim: core.Primitive, *args: TfVal, **kwargs) \
+    -> List[Limitation]:
+  """
+  Given a primitive and a set of parameters one would like to pass to it,
+  categorize identifies the potential limitations the call would encounter when
+  converted to TF through jax2tf.
+
+  Args:
+    prim: the primitive to call.
+    args: the arguments to pass to prim.
+    kwargs: the keyword arguments to pass to prim.
+
+  Returns:
+    A list of limitations
+  """
+  limitations: List[Limitation] = []
```

Comment:
Shouldn't this list be at a more global scope?

Reply:
My intention was to keep `categorize` self-contained; the list that is at a more global scope right now is `Categorizer().all_limitations`. If we decide that we want to simply populate a global list instead, then we can do that, but I was not really sold on this idea.

Reply:
I did not see where still is added to `all_limitations`